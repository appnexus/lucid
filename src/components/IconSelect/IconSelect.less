@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

@IconGroup_margin: 9px;
@iconBoxFontSize: 14px;

.@{prefix}-IconSelect {
	display: inline-flex;
	flex-wrap: wrap;

	> figure:not(:last-of-type) {
		margin: 0 @IconGroup_margin @IconGroup_margin 0;
	}
}

// Base Iconbox
.@{prefix}-IconSelect-Item {
	cursor: pointer;
	display: flex;
	flex-direction: column;
	justify-content: space-around;
	margin: 0;
	box-sizing: border-box;
	position: relative;
	align-items: center;
	padding: 12px 6px 11px;
	width: 120px;
	overflow: hidden;

	// Create border that doesn't effect layout
	&::before {
		content: '';
		top: 0;
		left: 0;
		position: absolute;
		border: 1px solid @color-gray;
		width: 100%;
		height: 100%;
		box-sizing: border-box;
		margin: 0 18px 0 0;
	}

	.@{prefix}-Icon {
		width: 72px;
		height: 32px;
		stroke: @color-neutral-9;
	}

	// Not Disabled
	&:not(.@{prefix}-IconSelect-Item-is-disabled):hover {
		cursor: pointer;

		&:not(.@{prefix}-IconSelect-Item-is-indeterminate)::before {
			border: 1px solid @color-primary;
		}

		.@{prefix}-Icon {
			opacity: 1;
			stroke: @color-primary;
		}

		.@{prefix}-Checkbox-visualization-container {
			border: 1px solid @color-primary;
		}
	}

	// On Hover, Active, Indeterminate, or Selected
	//&:not(.@{prefix}-IconSelect-Item-is-disabled):hover,
	&.@{prefix}-IconSelect-Item-is-active,
	&.@{prefix}-IconSelect-Item-is-selected {
		background-color: @color-primary-light;

		// on hover
		&:hover {
			background-color: @color-primary-light-hover;
		}

		&.@{prefix}-IconSelect-Item-is-disabled:hover {
			background-color: @color-primary-light;
		}

		&:not(.@{prefix}-IconSelect-Item-is-indeterminate)::before {
			border: 1px solid @color-primary;
		}

		.@{prefix}-Icon {
			opacity: 1;
			stroke: @color-primary;
		}

		.@{prefix}-Checkbox-visualization-container {
			border: 1px solid @color-primary;
		}
	}

	&.@{prefix}-IconSelect-Item-is-active,
	&:active {
		.@{prefix}-Checkbox-visualization-container {
			background: @color-primary;
		}
	}

	.@{prefix}-IconSelect-Item-figcaption,
	.@{prefix}-CheckboxLabeled,
	.@{prefix}-RadioButtonLabeled {
		font-size: @size-font;
		font-weight: @font-weight-medium;
		white-space: nowrap;
		text-align: center;
	}
}

// Disabled
.@{prefix}-IconSelect-Item-is-disabled {
	opacity: 0.9;

	&.@{prefix}-IconSelect-Item-is-active,
	&.@{prefix}-IconSelect-Item-is-selected {
		&.@{prefix}-IconSelect-Item-is-disabled::before {
			border: 1px solid @color-primary-50;
		}
	}

	.@{prefix}-CheckboxLabeled {
		color: @color-neutral-5;
	}

	&:hover {
		cursor: not-allowed;
	}

	.@{prefix}-Icon {
		stroke: @color-neutral-5;
	}
}

// Disabled && Selected
.@{prefix}-IconSelect-Item-is-disabled.@{prefix}-IconSelect-Item-is-selected {
	.@{prefix}-Icon {
		opacity: 0.5;
		stroke: @color-primary;
	}
}

// Has a Checkbox Input
.@{prefix}-IconSelect-Item-multi {
	height: 110px;
}

.@{prefix}-IconSelect-Item-checkbox {
	&.@{prefix}-CheckboxLabeled {
		display: flex;
		flex-direction: column;
		align-items: center;
	}

	&.@{prefix}-CheckboxLabeled > .@{prefix}-Checkbox {
		margin-right: 0;
	}

	.@{prefix}-CheckboxLabeled-Checkbox {
		order: 2;
	}

	.@{prefix}-CheckboxLabeled-label {
		order: 1;
		padding: 6px 0;
	}
}

// Has Radio Input
.@{prefix}-IconSelect-Item-single {
	height: 90px;
}

.@{prefix}-IconSelect-Item-radio {
	.@{prefix}-RadioButton.@{prefix}-IconSelect-Item-radio {
		visibility: hidden;
		position: absolute;
	}
}
