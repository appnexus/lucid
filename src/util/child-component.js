import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {

			/**
			 * getOwnChildren
			 *
			 * Returns all collective children of elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} All the collective children of elements with this component type
			 */
			getOwnChildren(elements) {
				let component = this;
				let children = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						children = children.concat(React.Children.map(element.props.children, x => x));
					}
				});
				return children;
			},

			/**
			 * getAll
			 *
			 * Returns all elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} elements with this component type
			 */
			getAll(elements) {
				let component = this;
				let elementsOfThisType = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						elementsOfThisType = elementsOfThisType.concat(element);
					}
				});
				return elementsOfThisType;
			},

			/**
			 * getAllProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} `props` objects of elements with this component type
			 */
			getAllProps(elements) {
				let component = this;
				return _.map(component.getAll(elements), 'props');
			},

			getAllAsProps(props, elements) {
				let component = this;
				return _.map(_.flatten([_.get(props, definition.propName, [])]), (childProp) => {
					if (_.isPlainObject(childProp)) {
						return _.assign({}, childProp, {children: childProp.children || childProp.text});
					} else {
						return {
							children: childProp
						};
					}
				}).concat(component.getAllProps(elements));
			}
		}
	});
}
