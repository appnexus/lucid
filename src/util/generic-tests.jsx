import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
