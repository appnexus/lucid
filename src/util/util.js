import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
