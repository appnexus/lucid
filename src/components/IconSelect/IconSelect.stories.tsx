import React, { useState } from 'react';
import { Meta } from '@storybook/react';
import IconSelect from './IconSelect';
import ClockIcon from '../Icon/ClockIcon/ClockIcon';
import StopwatchIcon from '../Icon/StopwatchIcon/StopwatchIcon';
import SwitchLabeled from '../SwitchLabeled/SwitchLabeled';

//👇 Provide Storybook with the component name, 'section', any subcomponents and a description
export default {
	title: 'Controls/IconSelect',
	component: IconSelect,
	parameters: {
		docs: {
			description: {
				component: IconSelect.peek.description,
			},
		},
	},
} as Meta;

//👇 Create a “template” of how args map to rendering
const Template: any = (args) => {
	const [selectedIcon, setSelectedIcon] = useState<string>('item2');

	const handleSelect = (id: string) => {
		// when selected, set `selectedIcon`
		setSelectedIcon(id);
	};

	return (
		<section>
			<IconSelect
				{...args}
				onSelect={handleSelect}
				items={
					[
						{
							id: 'item1',
							icon: <ClockIcon />,
							isSelected: selectedIcon === 'item1',
							label: 'Foo Bar',
						},
						{
							id: 'item2',
							icon: <StopwatchIcon />,
							isSelected: selectedIcon === 'item2',
							label: 'Bax Tar',
						},
					] as any
				}
			/>
		</section>
	);
};

//👇 Each story then reuses that template

/** Default: Select One Icon */
export const Default = Template.bind({});
Default.args = {
	kind: 'single', // renders as radio buttons
};

/** Select Multiple Icons */
export const SelectMultipleIcons = (args) => {
	const [selectedIcons, setSelectedIcons] = useState<string[]>(['item2']);

	const isSelected = (id: string) => {
		return selectedIcons.includes(id);
	};

	const handleSelect = (selectedId: string) => {
		if (isSelected(selectedId)) {
			// if selected, then remove from list
			setSelectedIcons(selectedIcons.filter((id: any) => id !== selectedId));
		} else {
			// add it to the list
			setSelectedIcons([...selectedIcons, selectedId]);
		}
	};

	return (
		<section>
			<IconSelect
				{...args}
				onSelect={handleSelect}
				items={
					[
						{
							id: 'item1',
							icon: <ClockIcon />,
							isSelected: isSelected('item1'),
							label: 'Foo Bar',
						},
						{
							id: 'item2',
							icon: <StopwatchIcon />,
							isSelected: isSelected('item2'),
							label: 'Bax Tar',
						},
					] as any
				}
			/>
		</section>
	);
};
SelectMultipleIcons.args = {
	kind: 'multiple', // default value, renders as checkboxes
};

/** Partially Select Icons */
export const PartiallySelectIcons = (args) => {
	const [selectedIcons, setSelectedIcons] = useState<any>([
		{ id: 'item2', isPartial: true },
	]);

	const isSelected = (id: any) => {
		//return false;
		return selectedIcons.some(({ id: selectedId }: any) => id === selectedId);
	};

	const isPartiallySelected = (id: any) => {
		//return false;
		return selectedIcons.some(
			({ id: selectedId, isPartial }: any) => id === selectedId && isPartial
		);
	};

	const handleSelect = (selectedId: any) => {
		//const icons = selectedIcons;

		if (isPartiallySelected(selectedId)) {
			// if partially selected, remove from the list
			setSelectedIcons(
				selectedIcons.filter(({ id }: any) => id !== selectedId)
			);
		} else if (isSelected(selectedId)) {
			// if already selected, ensure isPartial is true
			setSelectedIcons(
				selectedIcons.map(({ id, isPartial }: any) => ({
					id,
					isPartial: isPartial || id === selectedId,
				}))
			);
		} else {
			// add selected icon to list
			setSelectedIcons([
				...selectedIcons,
				{ id: selectedId, isPartial: false },
			]);
		}
	};

	return (
		<section>
			<IconSelect
				{...args}
				onSelect={handleSelect}
				items={[
					{
						id: 'item1',
						icon: <ClockIcon />,
						isSelected: isSelected('item1'),
						isPartial: isPartiallySelected('item1'),
						label: 'Foo Bar',
					},
					{
						id: 'item2',
						icon: <StopwatchIcon />,
						isSelected: isSelected('item2'),
						isPartial: isPartiallySelected('item2'),
						label: 'Bax Tar',
					},
				]}
			/>
		</section>
	);
};

/** Disabled Icon Select */
export const DisabledIconSelect = (args) => {
	const [selectedIcons, setSelectedIcons] = useState<string[]>(['item2']);
	const [isDisabled, setIsDisabled] = useState<boolean>(true);

	const isSelected = (id: any) => {
		return selectedIcons.includes(id);
	};

	const handleSelect = (selectedId: any) => {
		if (isSelected(selectedId)) {
			// if selected, then remove from the list
			setSelectedIcons(selectedIcons.filter((id: any) => id !== selectedId));
		} else {
			// add it to the list
			setSelectedIcons([...selectedIcons, selectedId]);
		}
	};

	const handleToggleDisabled = () => {
		setIsDisabled(!isDisabled);
	};

	return (
		<section>
			<SwitchLabeled
				Label={`IconSelect ${isDisabled ? 'Disabled' : 'Enabled'}`}
				isSelected={!isDisabled}
				onSelect={handleToggleDisabled}
			/>

			<IconSelect
				{...args}
				isDisabled={isDisabled}
				onSelect={handleSelect}
				items={
					[
						{
							id: 'item1',
							icon: <ClockIcon />,
							isSelected: isSelected('item1'),
							label: 'Always Disabled',
							isDisabled: true,
						},
						{
							id: 'item2',
							icon: <StopwatchIcon />,
							isSelected: isSelected('item2'),
							label: 'Bax Tar',
						},
					] as any
				}
			/>
		</section>
	);
};
