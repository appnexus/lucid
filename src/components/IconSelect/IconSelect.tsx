import _ from 'lodash';
import React from 'react';
import PropTypes from 'react-peek/prop-types';

import Icon from '../Icon/Icon';
import RadioButtonLabeled from '../RadioButtonLabeled/RadioButtonLabeled';
import CheckboxLabeled from '../CheckboxLabeled/CheckboxLabeled';

import { lucidClassNames } from '../../util/style-helpers';
import {
	StandardProps,
	omitProps,
	Overwrite,
} from '../../util/component-types';

const cx = lucidClassNames.bind('&-IconSelect');

const { arrayOf, bool, func, node, number, oneOf, string, shape } = PropTypes;

const getFigureParent = (domNode: HTMLElement): HTMLElement | undefined => {
	if (domNode.classList.contains(cx('&-Item'))) {
		return domNode;
	}
	if (domNode === document.body) {
		throw new Error(`domNode is not a child of .${cx('&-Item')}`);
	}
	if (domNode.parentElement) {
		return getFigureParent(domNode.parentElement);
	}

	return;
};

interface Item {
	id: string;
	icon?: React.ReactElement;
	label?: React.ReactElement;
	isSelected?: boolean;
	isPartial?: boolean;
	tabIndex?: number;
	isDisabled?: boolean;
	className?: string;
}

interface IIconSelectPropsRaw extends StandardProps {
	/** Items in the IconSelect group. Each item should have an id. */
	items: Item[];

	/** Defines the type of IconSelect box. A 'single' select will create a radio
		input type Item. A 'multiple' select will create a checkbox input type. */
	kind: 'single' | 'multiple';

	/** A function that is called with the id of the Item in the IconSelect group
		is clicked. */
	onSelect: (
		id: string,
		{
			event,
			props,
		}: {
			event: React.MouseEvent;
			props: IIconSelectProps;
		}
	) => void;

	/** Disabled all IconSelect Items. */
	isDisabled: boolean;
}

export type IIconSelectProps = Overwrite<
	React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLSpanElement>,
		HTMLSpanElement
	>,
	IIconSelectPropsRaw
>;

const defaultProps = {
	kind: 'multiple' as const,
	isDisabled: false,
	onSelect: _.noop,
};

export const IconSelect = (props: IIconSelectProps): React.ReactElement => {
	const {
		className,
		children,
		kind,
		items,
		isDisabled,
		onSelect,
		...passThroughs
	} = props;

	const handleClick = (event: React.MouseEvent): void => {
		if (!props.isDisabled) {
			const domNode = getFigureParent(event.target as HTMLElement);
			if (domNode) {
				const id = domNode.dataset.id;

				domNode.focus();
				if (!domNode.hasAttribute('disabled') && id) {
					onSelect(id, { event, props: props });
				}
			}
		}
	};

	const getChildIcon = (icon: React.ReactElement): React.ReactElement => {
		return icon ? (
			icon
		) : (
			<Icon>
				<rect x='0' y='0' width='16' height='16' />
				<rect x='1' y='1' width='14' height='14' fill='white' />
			</Icon>
		);
	};

	const getInputComponent = (item: Item) => {
		const { kind, className, isDisabled } = props;
		const Label = item.label;
		const singleSelect = _.isEqual(kind, 'single');

		return singleSelect ? (
			<RadioButtonLabeled
				Label={Label}
				className={cx('&-Item-radio', {
					[`${className}-radio`]: className,
				})}
				isDisabled={isDisabled || item.isDisabled}
				isSelected={item.isSelected}
				// tabIndex={item.tabIndex}
			/>
		) : (
			<CheckboxLabeled
				Label={Label}
				className={cx('&-Item-checkbox', {
					[`${className}-checkbox`]: className,
				})}
				isDisabled={isDisabled || item.isDisabled ? true : false}
				isIndeterminate={item.isPartial ? true : false}
				isSelected={item.isSelected ? true : false}
				// tabIndex={item.tabIndex}
			/>
		);
	};

	return (
		<span
			{...omitProps(passThroughs, undefined, [
				..._.keys(IconSelect.propTypes),
				'items',
			])}
			className={cx('&', className)}
		>
			{_.map(items, (childItem, index): React.ReactElement => {
				const itemDisabled = isDisabled || childItem.isDisabled;
				return (
					<figure
						key={`iconselectitem_${index}`}
						className={cx('&-Item', childItem.className, {
							[`${className}-Item`]: className,
							'&-Item-is-disabled': itemDisabled,
							'&-Item-is-partial': childItem.isPartial,
							'&-Item-is-selected': childItem.isSelected,
							'&-Item-multi': kind === 'multiple',
							'&-Item-single': kind === 'single',
						})}
						data-id={childItem.id}
						onClick={itemDisabled ? undefined : handleClick}
					>
						{childItem.icon && getChildIcon(childItem.icon)}
						<figcaption className={cx('&-Item-figcaption')}>
							{getInputComponent(childItem)}
						</figcaption>
					</figure>
				);
			})}
			{children}
		</span>
	);
};

IconSelect.displayName = 'IconSelect';

IconSelect.defaultProps = defaultProps;

IconSelect.peek = {
	description: `\`IconSelect\` allow you to pair icons together to form a related cluster. Any props not explicitly called out are spread on to the root component.
	`,
	categories: ['controls', 'selectors'],
};

IconSelect.propTypes = {
	className: string`
		Appended to the component-specific class names set on the root element.
		Value is run through the \`classnames\` library.
	`,

	children: node`
		Added to the end of the IconSelect group.
	`,

	items: arrayOf(
		shape({
			id: string.isRequired,
			icon: node,
			label: node,
			isSelected: bool,
			isPartial: bool,
			tabIndex: number,
			isDisabled: bool,
			className: string,
		})
	).isRequired`
		Items in the IconSelect group. Each item should have an id.
	`,

	kind: oneOf(['single', 'multiple'])`
		Defines the type of IconSelect box. A 'single' select will create a radio
		input type Item. A 'multiple' select will create a checkbox input type.
	`,

	onSelect: func`
		A function that is called with the id of the Item in the IconSelect group
		is clicked.  Signature: \`(id, { event }) => {}\`
	`,

	isDisabled: bool`
		Disabled all IconSelect Items.
	`,
};

export default IconSelect;
