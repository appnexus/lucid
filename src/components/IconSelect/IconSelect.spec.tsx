import React from 'react';
import assert from 'assert';
import { shallow } from 'enzyme';

import { common } from '../../util/generic-tests';
import IconSelect from './IconSelect';

const items = [
	{
		id: 'one',
		isSelected: false,
		label: 'one',
	},
	{
		id: 'two',
		isSelected: true,
		label: 'two',
	},
];

describe('IconSelect', () => {
	common(IconSelect, {
		defaultProps: { items: [] },
	});

	it('prop children', () => {
		const wrapper = shallow(
			<IconSelect items={items}>
				<div className='jim' />
			</IconSelect>
		);

		assert(wrapper.contains(<div className='jim' />));
	});

	it.skip('has a selected item', () => {
		const wrapper = shallow(<IconSelect items={items} />);

		assert.equal(
			wrapper
				.children()
				.at(1)
				.children()
				.at(1)
				.children()
				.prop('isSelected'),
			true
		);
	});

	describe('IconSelect Events', () => {
		it.skip('should call the onClick handler when clicked', () => {
			const onIconSelect = jest.fn();
			const onIconSelectClick = jest.fn();
			const mockEvent = {
				target: {
					classList: {
						contains: () => true,
					},
					focus: onIconSelect,
					hasAttribute: () => false,
					dataset: {
						id: 'test',
					},
				},
			};
			const wrapper = shallow(
				<IconSelect items={items} onSelect={onIconSelectClick} />
			);
			const onClickEvent = wrapper.instance();

			onClickEvent.handleClick(mockEvent);

			expect(onIconSelect).toHaveBeenCalled();
			expect(onIconSelectClick).toHaveBeenCalledWith(
				'test',
				expect.objectContaining({
					event: expect.anything(),
					props: expect.anything(),
				})
			);
		});

		it.skip('should not use onClick if disabled', () => {
			const onIconSelect = jest.fn();
			const onIconSelectClick = jest.fn();
			const mockEvent = {
				target: {
					focus: onIconSelect,
					hasAttribute: () => true,
					dataset: {
						id: 'test',
					},
				},
			};
			const wrapper = shallow(
				<IconSelect
					items={items}
					isDisabled={true}
					onSelect={onIconSelectClick}
				/>
			);
			const onClickEvent = wrapper.instance();

			onClickEvent.handleClick(mockEvent);

			expect(onIconSelect).not.toHaveBeenCalled();
			expect(onIconSelectClick).not.toHaveBeenCalled();
		});
	});
});
