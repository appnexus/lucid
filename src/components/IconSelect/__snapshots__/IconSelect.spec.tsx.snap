// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSelect [common] example testing should match snapshot(s) for 1.select-one-icon 1`] = `
<span
  className="lucid-IconSelect"
>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-single"
    data-id="item1"
    key="iconselectitem_0"
    onClick={[Function]}
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <RadioButtonLabeled
        Label="Foo Bar"
        className="lucid-IconSelect-Item-radio"
        isDisabled={false}
        isSelected={false}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-is-selected lucid-IconSelect-Item-single"
    data-id="item2"
    key="iconselectitem_1"
    onClick={[Function]}
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <RadioButtonLabeled
        Label="Bax Tar"
        className="lucid-IconSelect-Item-radio"
        isDisabled={false}
        isSelected={true}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
</span>
`;

exports[`IconSelect [common] example testing should match snapshot(s) for 2.select-multiple-icons 1`] = `
<span
  className="lucid-IconSelect"
>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-multi"
    data-id="item1"
    key="iconselectitem_0"
    onClick={[Function]}
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <CheckboxLabeled
        Label="Foo Bar"
        className="lucid-IconSelect-Item-checkbox"
        isDisabled={false}
        isIndeterminate={false}
        isSelected={false}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-is-selected lucid-IconSelect-Item-multi"
    data-id="item2"
    key="iconselectitem_1"
    onClick={[Function]}
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <CheckboxLabeled
        Label="Bax Tar"
        className="lucid-IconSelect-Item-checkbox"
        isDisabled={false}
        isIndeterminate={false}
        isSelected={true}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
</span>
`;

exports[`IconSelect [common] example testing should match snapshot(s) for 3.partially-select-icons 1`] = `
<span
  className="lucid-IconSelect"
>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-multi"
    data-id="item1"
    key="iconselectitem_0"
    onClick={[Function]}
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <CheckboxLabeled
        Label="Foo Bar"
        className="lucid-IconSelect-Item-checkbox"
        isDisabled={false}
        isIndeterminate={false}
        isSelected={false}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-is-partial lucid-IconSelect-Item-is-selected lucid-IconSelect-Item-multi"
    data-id="item2"
    key="iconselectitem_1"
    onClick={[Function]}
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <CheckboxLabeled
        Label="Bax Tar"
        className="lucid-IconSelect-Item-checkbox"
        isDisabled={false}
        isIndeterminate={true}
        isSelected={true}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
</span>
`;

exports[`IconSelect [common] example testing should match snapshot(s) for 4.disabled-icon-select 1`] = `
<span
  className="lucid-IconSelect"
>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-is-disabled lucid-IconSelect-Item-multi"
    data-id="item1"
    key="iconselectitem_0"
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <CheckboxLabeled
        Label="Always Disabled"
        className="lucid-IconSelect-Item-checkbox"
        isDisabled={true}
        isIndeterminate={false}
        isSelected={false}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
  <figure
    className="lucid-IconSelect-Item lucid-IconSelect-Item-is-disabled lucid-IconSelect-Item-is-selected lucid-IconSelect-Item-multi"
    data-id="item2"
    key="iconselectitem_1"
  >
    <ClockIcon
      aspectRatio="xMidYMid meet"
      color="primary"
      isClickable={false}
      isDisabled={false}
      onClick={[Function]}
      onSelect={[Function]}
      size={16}
      viewBox="0 0 16 16"
    />
    <figcaption
      className="lucid-IconSelect-Item-figcaption"
    >
      <CheckboxLabeled
        Label="Bax Tar"
        className="lucid-IconSelect-Item-checkbox"
        isDisabled={true}
        isIndeterminate={false}
        isSelected={true}
        onSelect={[Function]}
      />
    </figcaption>
  </figure>
</span>
`;
